client/finality-grandpa/src/tests.rs

```diff
@@ -1287,105 +1286,101 @@ fn voter_persists_its_votes() {
-						Ok(())
-					}).map_err(|_| panic!()))
-
-			} else if state.compare_and_swap(1, 2, Ordering::SeqCst) == 1 {
-				// the next message we receive should be our own prevote
-				let prevote = match signed.message {
-					finality_grandpa::Message::Prevote(prevote) => prevote,
-					_ => panic!("We should receive our own prevote."),
-				};
-
-				// targeting block 30
-				assert!(prevote.target_number == 30);
-
-				// after alice restarts it should send its previous prevote
-				// therefore we won't ever receive it again since it will be a
-				// known message on the gossip layer
-
-				future::Either::B(future::ok(()))
-
-			} else if state.compare_and_swap(2, 3, Ordering::SeqCst) == 2 {
-				// we then receive a precommit from alice for block 15
-				// even though we casted a prevote for block 30
-				let precommit = match signed.message {
-					finality_grandpa::Message::Precommit(precommit) => precommit,
-					_ => panic!("voter should precommit."),
-				};
-
-				assert!(precommit.target_number == 15);
-
-				// signal exit
-				exit_tx.clone().lock().take().unwrap().send(()).unwrap();
-
-				future::Either::B(future::ok(()))
+		let state = Arc::new(AtomicUsize::new(0));
+
+		runtime.spawn(round_rx.try_for_each(move |signed| {
+			let net2 = net.clone();
+			let net = net.clone();
+			let voter_tx = voter_tx.clone();
+			let round_tx = round_tx.clone();
+			let state = state.clone();
+			let exit_tx = exit_tx.clone();
+
+			async move {
+				if state.compare_and_swap(0, 1, Ordering::SeqCst) == 0 {
+					// the first message we receive should be a prevote from alice.
+					let prevote = match signed.message {
+						finality_grandpa::Message::Prevote(prevote) => prevote,
+						_ => panic!("voter should prevote."),
+					};
+
+					// its chain has 20 blocks and the voter targets 3/4 of the
+					// unfinalized chain, so the vote should be for block 15
+					assert!(prevote.target_number == 15);
+
+					// we push 20 more blocks to alice's chain
+					net.lock().peer(0).push_blocks(20, false);
+
+					let interval = futures::stream::unfold(Delay::new(Duration::from_millis(200)), |delay|
+						Box::pin(async move {
+							delay.await;
+							Some(((), Delay::new(Duration::from_millis(200))))
+						})
+					);
+
+					interval
+						.take_while(move |_| {
+							future::ready(net2.lock().peer(1).client().info().best_number != 40)
+						})
+						.for_each(|_| future::ready(()))
+						.await;
+
+					let block_30_hash =
+						net.lock().peer(0).client().as_full().unwrap().hash(30).unwrap().unwrap();
+
+					// we restart alice's voter
+					voter_tx.unbounded_send(()).unwrap();
+
+					// and we push our own prevote for block 30
+					let prevote = finality_grandpa::Prevote {
+						target_number: 30,
+						target_hash: block_30_hash,
+					};
+
+					Pin::new(&mut *round_tx.lock()).start_send(finality_grandpa::Message::Prevote(prevote)).unwrap();
```

Comment:
This leads to a compile error: 
```
error[E0308]: mismatched types
    --> client/finality-grandpa/src/tests.rs:1340:84
     |
1340 |                     Pin::new(&mut *round_tx.lock()).start_send(finality_grandpa::Message::Prevote(prevote)).unwrap();
     |                                                                                                   ^^^^^^^ expected associated type, found struct `primitive_types::H256`
     |
     = note: expected type `finality_grandpa::Prevote<<sp_runtime::generic::block::Block<sp_runtime::generic::header::Header<u64, sp_runtime::traits::BlakeTwo256>, substrate_test_runtime::Extrinsic> as sp_runtime::traits::Block>::Hash, <<sp_runtime::generic::block::Block<sp_runtime::generic::header::Header<u64, sp_runtime::traits::BlakeTwo256>, substrate_test_runtime::Extrinsic> as sp_runtime::traits::Block>::Header as sp_runtime::traits::Header>::Number>`
                found type `finality_grandpa::Prevote<primitive_types::H256, {integer}>`
     = note: consider constraining the associated type `<sp_runtime::generic::block::Block<sp_runtime::generic::header::Header<u64, sp_runtime::traits::BlakeTwo256>, substrate_test_runtime::Extrinsic> as sp_runtime::traits::Block>::Hash` to `primitive_types::H256` or calling a method that returns `<sp_runtime::generic::block::Block<sp_runtime::generic::header::Header<u64, sp_runtime::traits::BlakeTwo256>, substrate_test_runtime::Extrinsic> as sp_runtime::traits::Block>::Hash`
     = note: for more information, visit https://doc.rust-lang.org/book/ch19-03-advanced-traits.html

error: aborting due to previous error
```

Reply:
Try switching `round_communication` to return concrete type instead of anonymous `impl Sink`

Reply:
Yeah, it's probably unable to project through the anonymous return type (intended behavior - `impl Trait` should not be leaky)

Reply:
I confirmed that this works locally, but the behavior is definitely strange. It seems like `impl Trait<T>` can't unify projected versions of T with T.